zyte_common_items/pipelines.py

```diff
@@ -115,25 +116,44 @@ def get_threshold_for_item(self, item, spider):
     def get_item_name(self, item):
         return item.__class__.__name__
 
-    def process_item(self, item, spider):
-        from scrapy.exceptions import DropItem
-
-        item_name = self.get_item_name(item)
+    def _process_probability(self, item, threshold):
+        if not isinstance(item, ProbabilityMixin):
+            return True
         item_proba = item.get_probability()
-        threshold = self.get_threshold_for_item(item, spider)
-
+        if item_proba is None:
+            # don't emit stats for types without probability
+            return True
+        item_name = self.get_item_name(item)
         self.stats.inc_value("drop_low_probability_item/processed")
         self.stats.inc_value(f"drop_low_probability_item/processed/{item_name}")
-
-        if item_proba is None or item_proba >= threshold:
+        if item_proba >= threshold:
             self.stats.inc_value("drop_low_probability_item/kept")
             self.stats.inc_value(f"drop_low_probability_item/kept/{item_name}")
-            return item
-
+            return True
         self.stats.inc_value("drop_low_probability_item/dropped")
         self.stats.inc_value(f"drop_low_probability_item/dropped/{item_name}")
+        return False
+
+    def process_item(self, item, spider):
+        from scrapy.exceptions import DropItem
+
+        if isinstance(item, dict):
+            # for nested items remove sub-items that have low probability
+            # instead of dropping the whole result
+            new_item = {}
+            for item_type, sub_item in item.items():
+                threshold = self.get_threshold_for_item(sub_item, spider)
+                if self._process_probability(sub_item, threshold):
+                    new_item[item_type] = sub_item
+            if not new_item:
+                # everything has been removed
+                raise DropItem
```

Comment:
Hm, there is a difference in error messages - should we copy some of the text from the DropItem below? 
We also don't log anything when sub-items are dropped; probably that's ok, but maybe not - what do you think?

Reply:
+1 to logging something when dropping sub-items.